### PR TITLE
+str implement "bind completions" flow

### PR DESCRIPTION
**Pull request by ktoso**
Implements basic version of #21549.

Pending:
- decide where to put it. I like the naming of BindCompletion.inSomeDescriptiveWay
- adding docs

Implements just the basic one that forwards errors/completion/cancellation "all the way",
should already be useful for users who implement websockets using `Flow.fromSinkAndSource`

### CODE REVIEW

**Comment by drewhk** [comment]
ping @ktoso will you pick this up soon? Otherwise maybe stash it.

**Comment by ktoso** [comment]
Would we accept it? I was awaiting initial feedback. Happy to complete it if yes.

File: akka-stream-tests/src/test/scala/akka/stream/extra/CompletionExtraSpec.scala [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        override def onPush(): Unit = push(out1, grab(in1))
+        override def onPull(): Unit = pull(in1)
+      }
+
+      val handler2 = new InHandler with OutHandler {
+        override def onPush(): Unit = push(out2, grab(in2))
+        override def onPull(): Unit = pull(in2)
+      }
+
+      setHandlers(in1, out1, handler1)
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/extra/CompletionExtraSpec.scala]: passAlong?

**Review by drewhk — COMMENTED** [review]
looks like a good addition

**Comment by ktoso** [comment]
Note to self, complete everything once anyone completes:

```scala
  // FIXME, decide where to put it and how to expose
  class CoupledCompletionBidi[I, O] extends GraphStage[BidiShape[I, I, O, O]] {
    val in1: Inlet[I] = Inlet("CoupledCompletion.in1")
    val out1: Outlet[I] = Outlet("CoupledCompletion.out1")
    val in2: Inlet[O] = Inlet("CoupledCompletion.in2")
    val out2: Outlet[O] = Outlet("CoupledCompletion.out2")
    override val shape: BidiShape[I, I, O, O] = BidiShape(in1, out1, in2, out2)

    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {

      val handler1 = new InHandler with OutHandler {
        override def onPush(): Unit = push(out1, grab(in1))
        override def onPull(): Unit = pull(in1)

        override def onDownstreamFinish(): Unit = completeStage()
        override def onUpstreamFinish(): Unit = completeStage()
        override def onUpstreamFailure(ex: Throwable): Unit = failStage(ex)
      }

      val handler2 = new InHandler with OutHandler {
        override def onPush(): Unit = push(out2, grab(in2))
        override def onPull(): Unit = pull(in2)

        override def onDownstreamFinish(): Unit = completeStage()
        override def onUpstreamFinish(): Unit = completeStage()
        override def onUpstreamFailure(ex: Throwable): Unit = failStage(ex)
      }

      setHandlers(in1, out1, handler1)
      setHandlers(in2, out2, handler2)
    }
  }

```

**Comment by ktoso** [comment]
Now with docs and more tests, please review. @jrudolph @patriknw

**Review by johanandren — APPROVED** [review]
LGTM but lacks a mention in the big "all the stages" doc page.

**Comment by ktoso** [comment]
Added more docs, can I merge?

**Review by drewhk — APPROVED** [review]

**Pull request merged**